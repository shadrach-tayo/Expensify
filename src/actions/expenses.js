import database from '../firebase/firebase';

// addExpense
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const { uid } = getState().auth;
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, amount, note, createdAt };
    return database
      .ref(`users/${uid}/expenses`)
      .push(expense)
      .then(ref => {
        dispatch(addExpense({ id: ref.key, ...expense }));
      })
      .catch(err => {
        debugger;
      });
  };
};

// removeExpense
export const removeExpense = ({ id = '' } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

// startRemoveExpense
export const startRemoveExpense = ({ id = '' } = {}) => {
  return (dispatch, getState) => {
    const { uid } = getState().auth;
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense({ id }));
      });
  };
};

// editExpense
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

// startEditExpense
export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const { uid } = getState().auth;
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .update(updates)
      .then(() => dispatch(editExpense(id, updates)));
  };
};

export const setExpenses = expenses => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const { uid } = getState().auth;
    return database
      .ref(`users/${uid}/expenses`)
      .once('value')
      .then(snapshots => {
        const expenses = [];
        snapshots.forEach(snapshot => {
          expenses.push({ id: snapshot.key, ...snapshot.val() });
        });
        dispatch(setExpenses(expenses));
      });
  };
};
