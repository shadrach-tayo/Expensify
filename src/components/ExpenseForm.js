import React from 'react';
import { SingleDatePicker } from 'react-dates';
import moment from 'moment';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super();
    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      focused: false,
      error: ''
    };
  }

  onDescriptionchange = e => {
    this.setState({ description: e.target.value });
  };

  onNoteChange = e => {
    this.setState({ note: e.target.value });
  };

  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState({ amount: e.target.value });
    }
  };

  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({
        createdAt
      }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({
      focused
    }));
  };

  onSubmit = e => {
    e.preventDefault();
    const { description, amount, note, createdAt } = this.state;
    if (!description || !amount) {
      this.setState({ error: 'please fill in the description and the amount' });
    } else {
      this.setState({ error: '' });
      this.props.onSubmit({
        amount: parseInt(amount, 10) * 100,
        createdAt: createdAt.valueOf(),
        description,
        note
      });
    }
  };

  render() {
    return (
      <div className="container">
        <form className="form" onSubmit={this.onSubmit}>
          {this.state.error ? (
            <p className="form__error">{this.state.error}</p>
          ) : (
            ''
          )}
          <div className="input-group">
            <div className="input-group__item">
              <input
                type="text"
                className="text-input"
                placeholder="description"
                autoFocus
                value={this.state.description}
                onChange={this.onDescriptionchange}
              />
            </div>
            <div className="input-group__item">
              <SingleDatePicker
                date={this.state.createdAt}
                onDateChange={this.onDateChange}
                focused={this.state.focused}
                onFocusChange={this.onFocusChange}
                numberOfMonths={1}
                isOutsideRange={() => false}
              />
            </div>
            <div className="input-group__item">
              <textarea
                className="textarea"
                placeholder="Add expense note (optional)"
                value={this.state.note}
                onChange={this.onNoteChange}
              />
            </div>
            <div className="input-group__item">
              <input
                type="number"
                className="text-input"
                value={this.state.amount}
                placeholder="amount"
                onChange={this.onAmountChange}
              />
            </div>
            <div className="input-group__item">
              <button className="btn">Save Expense</button>
            </div>
          </div>
        </form>
      </div>
    );
  }
}
