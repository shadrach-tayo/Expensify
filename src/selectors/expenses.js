import moment from 'moment';

// Get Visible Expenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt);
      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());
      const startDateMatch = startDate
        ? moment(startDate).isSameOrBefore(createdAtMoment)
        : true;
      const endDateMatch = endDate
        ? moment(endDate).isSameOrAfter(createdAtMoment)
        : true;
      return textMatch && startDateMatch && endDateMatch;
    })
    .sort((a, b) => {
      if (sortBy === 'date') return a.createdAt < b.createdAt ? 1 : -1;
      if (sortBy === 'amount') return a.amount < b.amount ? 1 : -1;
    });
};
