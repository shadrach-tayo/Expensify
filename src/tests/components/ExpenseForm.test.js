import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import moment from 'moment';
import ExpenseForm from '../../components/ExpenseForm';
import expenses from '../fixtures/expenses';

test('should render ExpenseForm correctly', () => {
  const wrapper = shallow(<ExpenseForm />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should render ExpenseForm correctly with expense data', () => {
  const wrapper = shallow(<ExpenseForm expense={expenses[0]} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should render error with invalid form submission', () => {
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {}
  });
  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot();
});

test('should set description on input change', () => {
  const value = 'New Description';
  const wrapper = shallow(<ExpenseForm />);
  wrapper
    .find('input')
    .at(0)
    .simulate('change', {
      target: { value }
    });
  expect(wrapper.state('description')).toBe(value);
  expect(wrapper).toMatchSnapshot();
});

test('should set note on textarea change', () => {
  const value = 'New Note';
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find('textarea').simulate('change', {
    target: { value }
  });
  expect(wrapper.state('note')).toBe(value);
  expect(wrapper).toMatchSnapshot();
});

test('should set amount if valid input', () => {
  const value = '23.5';
  const wrapper = shallow(<ExpenseForm />);
  wrapper
    .find('input')
    .at(1)
    .simulate('change', {
      target: { value }
    });
  expect(wrapper.state('amount')).toBe(value);
  expect(wrapper).toMatchSnapshot();
});

test('should not set amount if invalid input', () => {
  const value = '12.122';
  const wrapper = shallow(<ExpenseForm />);
  wrapper
    .find('input')
    .at(1)
    .simulate('change', {
      target: { value }
    });
  expect(wrapper.state('amount')).toBe('');
  expect(wrapper).toMatchSnapshot();
});

test('should call onSubmit prop for valid form submission', () => {
  const onSubmitSpy = jest.fn();
  const wrapper = shallow(
    <ExpenseForm expense={expenses[0]} onSubmit={onSubmitSpy} />
  );
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {}
  });
  expect(wrapper.state('error')).toBe('');
  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    description: expenses[0].description,
    amount: expenses[0].amount,
    note: expenses[0].note,
    createdAt: expenses[0].createdAt
  });
});

test('should set new date on date change', () => {
  const wrapper = shallow(<ExpenseForm />);
  const now = moment();
  wrapper.find('SingleDatePicker').prop('onDateChange')(now);
  expect(wrapper.state('createdAt')).toEqual(now);
});

test('should set new calender focus on change', () => {
  const wrapper = shallow(<ExpenseForm />);
  const focused = true;
  wrapper.find('SingleDatePicker').prop('onFocusChange')({ focused });
  expect(wrapper.state('focused')).toEqual(focused);
});
