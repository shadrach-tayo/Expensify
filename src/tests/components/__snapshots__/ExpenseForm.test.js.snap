// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not set amount if invalid input 1`] = `
<div
  className="container"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="input-group"
    >
      <div
        className="input-group__item"
      >
        <input
          autoFocus={true}
          className="text-input"
          onChange={[Function]}
          placeholder="description"
          type="text"
          value=""
        />
      </div>
      <div
        className="input-group__item"
      >
        <SingleDatePicker
          anchorDirection="left"
          customCloseIcon={null}
          customInputIcon={null}
          date={"1970-01-01T00:00:00.000Z"}
          daySize={39}
          disabled={false}
          displayFormat={[Function]}
          enableOutsideDays={false}
          firstDayOfWeek={null}
          focused={false}
          hideKeyboardShortcutsPanel={false}
          horizontalMargin={0}
          id="date"
          initialVisibleMonth={null}
          isDayBlocked={[Function]}
          isDayHighlighted={[Function]}
          isOutsideRange={[Function]}
          isRTL={false}
          keepOpenOnDateSelect={false}
          monthFormat="MMMM YYYY"
          navNext={null}
          navPrev={null}
          numberOfMonths={1}
          onClose={[Function]}
          onDateChange={[Function]}
          onFocusChange={[Function]}
          onNextMonthClick={[Function]}
          onPrevMonthClick={[Function]}
          orientation="horizontal"
          phrases={
            Object {
              "chooseAvailableDate": [Function],
              "clearDate": "Clear Date",
              "closeDatePicker": "Close",
              "dateIsUnavailable": [Function],
              "enterKey": "Enter key",
              "escape": "Escape key",
              "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
              "homeEnd": "Home and end keys",
              "jumpToNextMonth": "Move forward to switch to the next month",
              "jumpToPrevMonth": "Move backward to switch to the previous month",
              "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
              select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
              "keyboardShortcuts": "Keyboard Shortcuts",
              "leftArrowRightArrow": "Right and left arrow keys",
              "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
              "moveFocusByOneMonth": "Switch months",
              "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
              "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
              "openThisPanel": "Open this panel",
              "pageUpPageDown": "page up and page down keys",
              "questionMark": "Question mark",
              "returnFocusToInput": "Return to the date input field",
              "selectFocusedDate": "Select the date in focus",
              "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
              "upArrowDownArrow": "up and down arrow keys",
            }
          }
          placeholder="Date"
          readOnly={false}
          renderCalendarInfo={null}
          renderDay={null}
          renderMonth={null}
          reopenPickerOnClearDate={false}
          required={false}
          screenReaderInputMessage=""
          showClearDate={false}
          showDefaultInputIcon={false}
          withFullScreenPortal={false}
          withPortal={false}
        />
      </div>
      <div
        className="input-group__item"
      >
        <textarea
          className="textarea"
          onChange={[Function]}
          placeholder="Add expense note (optional)"
          value=""
        />
      </div>
      <div
        className="input-group__item"
      >
        <input
          className="text-input"
          onChange={[Function]}
          placeholder="amount"
          type="number"
          value=""
        />
      </div>
      <div
        className="input-group__item"
      >
        <button
          className="btn"
        >
          Save Expense
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`should render ExpenseForm correctly 1`] = `
<div
  className="container"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="input-group"
    >
      <div
        className="input-group__item"
      >
        <input
          autoFocus={true}
          className="text-input"
          onChange={[Function]}
          placeholder="description"
          type="text"
          value=""
        />
      </div>
      <div
        className="input-group__item"
      >
        <SingleDatePicker
          anchorDirection="left"
          customCloseIcon={null}
          customInputIcon={null}
          date={"1970-01-01T00:00:00.000Z"}
          daySize={39}
          disabled={false}
          displayFormat={[Function]}
          enableOutsideDays={false}
          firstDayOfWeek={null}
          focused={false}
          hideKeyboardShortcutsPanel={false}
          horizontalMargin={0}
          id="date"
          initialVisibleMonth={null}
          isDayBlocked={[Function]}
          isDayHighlighted={[Function]}
          isOutsideRange={[Function]}
          isRTL={false}
          keepOpenOnDateSelect={false}
          monthFormat="MMMM YYYY"
          navNext={null}
          navPrev={null}
          numberOfMonths={1}
          onClose={[Function]}
          onDateChange={[Function]}
          onFocusChange={[Function]}
          onNextMonthClick={[Function]}
          onPrevMonthClick={[Function]}
          orientation="horizontal"
          phrases={
            Object {
              "chooseAvailableDate": [Function],
              "clearDate": "Clear Date",
              "closeDatePicker": "Close",
              "dateIsUnavailable": [Function],
              "enterKey": "Enter key",
              "escape": "Escape key",
              "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
              "homeEnd": "Home and end keys",
              "jumpToNextMonth": "Move forward to switch to the next month",
              "jumpToPrevMonth": "Move backward to switch to the previous month",
              "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
              select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
              "keyboardShortcuts": "Keyboard Shortcuts",
              "leftArrowRightArrow": "Right and left arrow keys",
              "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
              "moveFocusByOneMonth": "Switch months",
              "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
              "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
              "openThisPanel": "Open this panel",
              "pageUpPageDown": "page up and page down keys",
              "questionMark": "Question mark",
              "returnFocusToInput": "Return to the date input field",
              "selectFocusedDate": "Select the date in focus",
              "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
              "upArrowDownArrow": "up and down arrow keys",
            }
          }
          placeholder="Date"
          readOnly={false}
          renderCalendarInfo={null}
          renderDay={null}
          renderMonth={null}
          reopenPickerOnClearDate={false}
          required={false}
          screenReaderInputMessage=""
          showClearDate={false}
          showDefaultInputIcon={false}
          withFullScreenPortal={false}
          withPortal={false}
        />
      </div>
      <div
        className="input-group__item"
      >
        <textarea
          className="textarea"
          onChange={[Function]}
          placeholder="Add expense note (optional)"
          value=""
        />
      </div>
      <div
        className="input-group__item"
      >
        <input
          className="text-input"
          onChange={[Function]}
          placeholder="amount"
          type="number"
          value=""
        />
      </div>
      <div
        className="input-group__item"
      >
        <button
          className="btn"
        >
          Save Expense
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`should render ExpenseForm correctly with expense data 1`] = `
<div
  className="container"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="input-group"
    >
      <div
        className="input-group__item"
      >
        <input
          autoFocus={true}
          className="text-input"
          onChange={[Function]}
          placeholder="description"
          type="text"
          value="new expense"
        />
      </div>
      <div
        className="input-group__item"
      >
        <SingleDatePicker
          anchorDirection="left"
          customCloseIcon={null}
          customInputIcon={null}
          date={"1970-01-01T00:00:00.000Z"}
          daySize={39}
          disabled={false}
          displayFormat={[Function]}
          enableOutsideDays={false}
          firstDayOfWeek={null}
          focused={false}
          hideKeyboardShortcutsPanel={false}
          horizontalMargin={0}
          id="date"
          initialVisibleMonth={null}
          isDayBlocked={[Function]}
          isDayHighlighted={[Function]}
          isOutsideRange={[Function]}
          isRTL={false}
          keepOpenOnDateSelect={false}
          monthFormat="MMMM YYYY"
          navNext={null}
          navPrev={null}
          numberOfMonths={1}
          onClose={[Function]}
          onDateChange={[Function]}
          onFocusChange={[Function]}
          onNextMonthClick={[Function]}
          onPrevMonthClick={[Function]}
          orientation="horizontal"
          phrases={
            Object {
              "chooseAvailableDate": [Function],
              "clearDate": "Clear Date",
              "closeDatePicker": "Close",
              "dateIsUnavailable": [Function],
              "enterKey": "Enter key",
              "escape": "Escape key",
              "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
              "homeEnd": "Home and end keys",
              "jumpToNextMonth": "Move forward to switch to the next month",
              "jumpToPrevMonth": "Move backward to switch to the previous month",
              "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
              select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
              "keyboardShortcuts": "Keyboard Shortcuts",
              "leftArrowRightArrow": "Right and left arrow keys",
              "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
              "moveFocusByOneMonth": "Switch months",
              "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
              "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
              "openThisPanel": "Open this panel",
              "pageUpPageDown": "page up and page down keys",
              "questionMark": "Question mark",
              "returnFocusToInput": "Return to the date input field",
              "selectFocusedDate": "Select the date in focus",
              "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
              "upArrowDownArrow": "up and down arrow keys",
            }
          }
          placeholder="Date"
          readOnly={false}
          renderCalendarInfo={null}
          renderDay={null}
          renderMonth={null}
          reopenPickerOnClearDate={false}
          required={false}
          screenReaderInputMessage=""
          showClearDate={false}
          showDefaultInputIcon={false}
          withFullScreenPortal={false}
          withPortal={false}
        />
      </div>
      <div
        className="input-group__item"
      >
        <textarea
          className="textarea"
          onChange={[Function]}
          placeholder="Add expense note (optional)"
          value=""
        />
      </div>
      <div
        className="input-group__item"
      >
        <input
          className="text-input"
          onChange={[Function]}
          placeholder="amount"
          type="number"
          value="0"
        />
      </div>
      <div
        className="input-group__item"
      >
        <button
          className="btn"
        >
          Save Expense
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`should render error with invalid form submission 1`] = `
<div
  className="container"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <p
      className="form__error"
    >
      please fill in the description and the amount
    </p>
    <div
      className="input-group"
    >
      <div
        className="input-group__item"
      >
        <input
          autoFocus={true}
          className="text-input"
          onChange={[Function]}
          placeholder="description"
          type="text"
          value=""
        />
      </div>
      <div
        className="input-group__item"
      >
        <SingleDatePicker
          anchorDirection="left"
          customCloseIcon={null}
          customInputIcon={null}
          date={"1970-01-01T00:00:00.000Z"}
          daySize={39}
          disabled={false}
          displayFormat={[Function]}
          enableOutsideDays={false}
          firstDayOfWeek={null}
          focused={false}
          hideKeyboardShortcutsPanel={false}
          horizontalMargin={0}
          id="date"
          initialVisibleMonth={null}
          isDayBlocked={[Function]}
          isDayHighlighted={[Function]}
          isOutsideRange={[Function]}
          isRTL={false}
          keepOpenOnDateSelect={false}
          monthFormat="MMMM YYYY"
          navNext={null}
          navPrev={null}
          numberOfMonths={1}
          onClose={[Function]}
          onDateChange={[Function]}
          onFocusChange={[Function]}
          onNextMonthClick={[Function]}
          onPrevMonthClick={[Function]}
          orientation="horizontal"
          phrases={
            Object {
              "chooseAvailableDate": [Function],
              "clearDate": "Clear Date",
              "closeDatePicker": "Close",
              "dateIsUnavailable": [Function],
              "enterKey": "Enter key",
              "escape": "Escape key",
              "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
              "homeEnd": "Home and end keys",
              "jumpToNextMonth": "Move forward to switch to the next month",
              "jumpToPrevMonth": "Move backward to switch to the previous month",
              "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
              select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
              "keyboardShortcuts": "Keyboard Shortcuts",
              "leftArrowRightArrow": "Right and left arrow keys",
              "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
              "moveFocusByOneMonth": "Switch months",
              "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
              "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
              "openThisPanel": "Open this panel",
              "pageUpPageDown": "page up and page down keys",
              "questionMark": "Question mark",
              "returnFocusToInput": "Return to the date input field",
              "selectFocusedDate": "Select the date in focus",
              "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
              "upArrowDownArrow": "up and down arrow keys",
            }
          }
          placeholder="Date"
          readOnly={false}
          renderCalendarInfo={null}
          renderDay={null}
          renderMonth={null}
          reopenPickerOnClearDate={false}
          required={false}
          screenReaderInputMessage=""
          showClearDate={false}
          showDefaultInputIcon={false}
          withFullScreenPortal={false}
          withPortal={false}
        />
      </div>
      <div
        className="input-group__item"
      >
        <textarea
          className="textarea"
          onChange={[Function]}
          placeholder="Add expense note (optional)"
          value=""
        />
      </div>
      <div
        className="input-group__item"
      >
        <input
          className="text-input"
          onChange={[Function]}
          placeholder="amount"
          type="number"
          value=""
        />
      </div>
      <div
        className="input-group__item"
      >
        <button
          className="btn"
        >
          Save Expense
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`should set amount if valid input 1`] = `
<div
  className="container"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="input-group"
    >
      <div
        className="input-group__item"
      >
        <input
          autoFocus={true}
          className="text-input"
          onChange={[Function]}
          placeholder="description"
          type="text"
          value=""
        />
      </div>
      <div
        className="input-group__item"
      >
        <SingleDatePicker
          anchorDirection="left"
          customCloseIcon={null}
          customInputIcon={null}
          date={"1970-01-01T00:00:00.000Z"}
          daySize={39}
          disabled={false}
          displayFormat={[Function]}
          enableOutsideDays={false}
          firstDayOfWeek={null}
          focused={false}
          hideKeyboardShortcutsPanel={false}
          horizontalMargin={0}
          id="date"
          initialVisibleMonth={null}
          isDayBlocked={[Function]}
          isDayHighlighted={[Function]}
          isOutsideRange={[Function]}
          isRTL={false}
          keepOpenOnDateSelect={false}
          monthFormat="MMMM YYYY"
          navNext={null}
          navPrev={null}
          numberOfMonths={1}
          onClose={[Function]}
          onDateChange={[Function]}
          onFocusChange={[Function]}
          onNextMonthClick={[Function]}
          onPrevMonthClick={[Function]}
          orientation="horizontal"
          phrases={
            Object {
              "chooseAvailableDate": [Function],
              "clearDate": "Clear Date",
              "closeDatePicker": "Close",
              "dateIsUnavailable": [Function],
              "enterKey": "Enter key",
              "escape": "Escape key",
              "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
              "homeEnd": "Home and end keys",
              "jumpToNextMonth": "Move forward to switch to the next month",
              "jumpToPrevMonth": "Move backward to switch to the previous month",
              "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
              select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
              "keyboardShortcuts": "Keyboard Shortcuts",
              "leftArrowRightArrow": "Right and left arrow keys",
              "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
              "moveFocusByOneMonth": "Switch months",
              "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
              "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
              "openThisPanel": "Open this panel",
              "pageUpPageDown": "page up and page down keys",
              "questionMark": "Question mark",
              "returnFocusToInput": "Return to the date input field",
              "selectFocusedDate": "Select the date in focus",
              "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
              "upArrowDownArrow": "up and down arrow keys",
            }
          }
          placeholder="Date"
          readOnly={false}
          renderCalendarInfo={null}
          renderDay={null}
          renderMonth={null}
          reopenPickerOnClearDate={false}
          required={false}
          screenReaderInputMessage=""
          showClearDate={false}
          showDefaultInputIcon={false}
          withFullScreenPortal={false}
          withPortal={false}
        />
      </div>
      <div
        className="input-group__item"
      >
        <textarea
          className="textarea"
          onChange={[Function]}
          placeholder="Add expense note (optional)"
          value=""
        />
      </div>
      <div
        className="input-group__item"
      >
        <input
          className="text-input"
          onChange={[Function]}
          placeholder="amount"
          type="number"
          value="23.5"
        />
      </div>
      <div
        className="input-group__item"
      >
        <button
          className="btn"
        >
          Save Expense
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`should set description on input change 1`] = `
<div
  className="container"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="input-group"
    >
      <div
        className="input-group__item"
      >
        <input
          autoFocus={true}
          className="text-input"
          onChange={[Function]}
          placeholder="description"
          type="text"
          value="New Description"
        />
      </div>
      <div
        className="input-group__item"
      >
        <SingleDatePicker
          anchorDirection="left"
          customCloseIcon={null}
          customInputIcon={null}
          date={"1970-01-01T00:00:00.000Z"}
          daySize={39}
          disabled={false}
          displayFormat={[Function]}
          enableOutsideDays={false}
          firstDayOfWeek={null}
          focused={false}
          hideKeyboardShortcutsPanel={false}
          horizontalMargin={0}
          id="date"
          initialVisibleMonth={null}
          isDayBlocked={[Function]}
          isDayHighlighted={[Function]}
          isOutsideRange={[Function]}
          isRTL={false}
          keepOpenOnDateSelect={false}
          monthFormat="MMMM YYYY"
          navNext={null}
          navPrev={null}
          numberOfMonths={1}
          onClose={[Function]}
          onDateChange={[Function]}
          onFocusChange={[Function]}
          onNextMonthClick={[Function]}
          onPrevMonthClick={[Function]}
          orientation="horizontal"
          phrases={
            Object {
              "chooseAvailableDate": [Function],
              "clearDate": "Clear Date",
              "closeDatePicker": "Close",
              "dateIsUnavailable": [Function],
              "enterKey": "Enter key",
              "escape": "Escape key",
              "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
              "homeEnd": "Home and end keys",
              "jumpToNextMonth": "Move forward to switch to the next month",
              "jumpToPrevMonth": "Move backward to switch to the previous month",
              "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
              select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
              "keyboardShortcuts": "Keyboard Shortcuts",
              "leftArrowRightArrow": "Right and left arrow keys",
              "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
              "moveFocusByOneMonth": "Switch months",
              "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
              "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
              "openThisPanel": "Open this panel",
              "pageUpPageDown": "page up and page down keys",
              "questionMark": "Question mark",
              "returnFocusToInput": "Return to the date input field",
              "selectFocusedDate": "Select the date in focus",
              "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
              "upArrowDownArrow": "up and down arrow keys",
            }
          }
          placeholder="Date"
          readOnly={false}
          renderCalendarInfo={null}
          renderDay={null}
          renderMonth={null}
          reopenPickerOnClearDate={false}
          required={false}
          screenReaderInputMessage=""
          showClearDate={false}
          showDefaultInputIcon={false}
          withFullScreenPortal={false}
          withPortal={false}
        />
      </div>
      <div
        className="input-group__item"
      >
        <textarea
          className="textarea"
          onChange={[Function]}
          placeholder="Add expense note (optional)"
          value=""
        />
      </div>
      <div
        className="input-group__item"
      >
        <input
          className="text-input"
          onChange={[Function]}
          placeholder="amount"
          type="number"
          value=""
        />
      </div>
      <div
        className="input-group__item"
      >
        <button
          className="btn"
        >
          Save Expense
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`should set note on textarea change 1`] = `
<div
  className="container"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="input-group"
    >
      <div
        className="input-group__item"
      >
        <input
          autoFocus={true}
          className="text-input"
          onChange={[Function]}
          placeholder="description"
          type="text"
          value=""
        />
      </div>
      <div
        className="input-group__item"
      >
        <SingleDatePicker
          anchorDirection="left"
          customCloseIcon={null}
          customInputIcon={null}
          date={"1970-01-01T00:00:00.000Z"}
          daySize={39}
          disabled={false}
          displayFormat={[Function]}
          enableOutsideDays={false}
          firstDayOfWeek={null}
          focused={false}
          hideKeyboardShortcutsPanel={false}
          horizontalMargin={0}
          id="date"
          initialVisibleMonth={null}
          isDayBlocked={[Function]}
          isDayHighlighted={[Function]}
          isOutsideRange={[Function]}
          isRTL={false}
          keepOpenOnDateSelect={false}
          monthFormat="MMMM YYYY"
          navNext={null}
          navPrev={null}
          numberOfMonths={1}
          onClose={[Function]}
          onDateChange={[Function]}
          onFocusChange={[Function]}
          onNextMonthClick={[Function]}
          onPrevMonthClick={[Function]}
          orientation="horizontal"
          phrases={
            Object {
              "chooseAvailableDate": [Function],
              "clearDate": "Clear Date",
              "closeDatePicker": "Close",
              "dateIsUnavailable": [Function],
              "enterKey": "Enter key",
              "escape": "Escape key",
              "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
              "homeEnd": "Home and end keys",
              "jumpToNextMonth": "Move forward to switch to the next month",
              "jumpToPrevMonth": "Move backward to switch to the previous month",
              "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
              select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
              "keyboardShortcuts": "Keyboard Shortcuts",
              "leftArrowRightArrow": "Right and left arrow keys",
              "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
              "moveFocusByOneMonth": "Switch months",
              "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
              "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
              "openThisPanel": "Open this panel",
              "pageUpPageDown": "page up and page down keys",
              "questionMark": "Question mark",
              "returnFocusToInput": "Return to the date input field",
              "selectFocusedDate": "Select the date in focus",
              "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
              "upArrowDownArrow": "up and down arrow keys",
            }
          }
          placeholder="Date"
          readOnly={false}
          renderCalendarInfo={null}
          renderDay={null}
          renderMonth={null}
          reopenPickerOnClearDate={false}
          required={false}
          screenReaderInputMessage=""
          showClearDate={false}
          showDefaultInputIcon={false}
          withFullScreenPortal={false}
          withPortal={false}
        />
      </div>
      <div
        className="input-group__item"
      >
        <textarea
          className="textarea"
          onChange={[Function]}
          placeholder="Add expense note (optional)"
          value="New Note"
        />
      </div>
      <div
        className="input-group__item"
      >
        <input
          className="text-input"
          onChange={[Function]}
          placeholder="amount"
          type="number"
          value=""
        />
      </div>
      <div
        className="input-group__item"
      >
        <button
          className="btn"
        >
          Save Expense
        </button>
      </div>
    </div>
  </form>
</div>
`;
